fearless/ApplicationLayer/Services/WalletBalanceSubscriptionAdapter.swift

```diff
@@ -0,0 +1,470 @@
+                return
+            }
+            guard let chainsResult = chainsOperation.result else {
+                self?.deliverOnCompletion(.failure(WalletBalanceError.chainsMissing))
+                return
+            }
+
+            switch (metaAccountResult, chainsResult) {
+            case let (.success(wallet), .success(chains)):
+                guard let wallet = wallet else {
+                    self?.deliverOnCompletion(.failure(WalletBalanceError.accountMissing))
+                    return
+                }
+                self?.handle([wallet], chains)
+            case let (.failure(error), _):
+                self?.deliverOnCompletion(.failure(error))
+            case let (_, .failure(error)):
+                self?.deliverOnCompletion(.failure(error))
+            }
+        }
+
+        operationQueue.addOperations([chainsOperation, metaAccountOperation], waitUntilFinished: false)
+    }
+
+    private func fetchAllMetaAccounts() {
+        let metaAccountsOperation = metaAccountRepository.fetchAllOperation(with: RepositoryFetchOptions())
+        let chainsOperation = fetchChainsOperation()
+
+        metaAccountsOperation.addDependency(chainsOperation)
+
+        metaAccountsOperation.completionBlock = { [weak self] in
+            guard let metaAccountsResult = metaAccountsOperation.result else {
+                self?.deliverOnCompletion(.failure(WalletBalanceError.accountMissing))
+                return
+            }
+            guard let chainsResult = chainsOperation.result else {
+                self?.deliverOnCompletion(.failure(WalletBalanceError.chainsMissing))
+                return
+            }
+
+            switch (metaAccountsResult, chainsResult) {
+            case let (.success(wallets), .success(chains)):
+                self?.handle(wallets, chains)
+            case let (.failure(error), _):
+                self?.deliverOnCompletion(.failure(error))
+            case let (_, .failure(error)):
+                self?.deliverOnCompletion(.failure(error))
+            }
+        }
+
+        operationQueue.addOperations([chainsOperation, metaAccountsOperation], waitUntilFinished: false)
+    }
+
+    private func subscribeToAccountInfo(
+        for wallets: [MetaAccountModel],
+        _ chainAssets: [ChainAsset]
+    ) {
+        wallets.forEach { wallet in
+            let accountInfoSubscriptionAdapter = AccountInfoSubscriptionAdapter(
+                walletLocalSubscriptionFactory: WalletLocalSubscriptionFactory.shared,
+                selectedMetaAccount: wallet
+            )
+
+            accountInfosAdapters[wallet.identifier] = accountInfoSubscriptionAdapter
+            accountInfoSubscriptionAdapter.subscribe(chainsAssets: chainAssets, handler: self)
+        }
+    }
+
+    private func subscribeToPrices(for chains: [ChainModel]) {
+        let pricesIds = chains.compactMap { $0.assets.compactMap(\.asset.priceId) }.reduce([], +)
```

Comment:
can't you pass [ChainAsset] array? two reasons to do this:
1. it will be unified with method above (subscribeToAccountInfo)
2. map func will be shorter and more beautiful
`chainAssets.map {$0.asset.priceId}`